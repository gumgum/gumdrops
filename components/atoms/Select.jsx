import React from 'react';
import PropTypes from 'prop-types';
import trimString from '../utils/trimString';

const Select = ({ context, size, customValue, customName, options, className, ...otherProps }) => {
    const baseClass = 'gds-form-group__select-input',
        contextClass = context ? `${baseClass}--${context}` : '',
        sizeClass = size ? `${baseClass}--${size}` : '';

    const classNames = trimString(`${baseClass} ${contextClass} ${sizeClass} ${className}`);

    return (
        <select className={classNames} {...otherProps}>
            {options.map(o => (
                <option key={o[customValue]} value={o[customValue]}>
                    {o[customName]}
                </option>
            ))}
        </select>
    );
};

Select.displayName = 'Select';

Select.defaultProps = {
    options: [],
    customValue: 'value',
    customName: 'name'
};

Select.propTypes = {
    /** primary, no-border, dark */
    context: PropTypes.string,
    /** xs, sm, lg */
    size: PropTypes.oneOf(['xs', 'sm', 'lg']),
    customValue: PropTypes.string,
    customName: PropTypes.string,
    options: PropTypes.array,
    className: PropTypes.string
};

export default Select;
