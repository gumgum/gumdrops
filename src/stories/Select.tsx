import React, { ChangeEventHandler } from 'react';
import cx from 'classnames';
import { Colors, SelectOptions, Sizes } from 'types';

export interface SelectPropsInterface extends React.SelectHTMLAttributes<HTMLSelectElement> {
    onChange: (value: unknown) => void;
    options: SelectOptions;
    color?: Colors;
    customValue?: string;
    customName?: string;
}

export type SelectProps = Omit<SelectPropsInterface, 'size'> & {
    size?: Sizes;
};

const baseClass = 'gds-form-group__select-input';

export const Select: React.FC<SelectProps> = ({
    onChange,
    color,
    size,
    customValue = 'value',
    customName = 'name',
    options,
    className,
    disabled,
    ...otherProps
}) => {
    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${color}`]: color,
        [`${baseClass}--${size}`]: size,
        '-disabled': disabled,
        '-pointer-events--none': disabled
    });

    const getChange: ChangeEventHandler = event => {
        const value = (event.currentTarget as HTMLInputElement).value;
        onChange(value);
    };

    return (
        <select onChange={getChange} className={rootClass} disabled={disabled} {...otherProps}>
            {options.map(option => {
                const isObject = typeof option === 'object' && option !== null;
                const key = isObject ? option[customName] : option;
                const value = isObject ? option[customValue] : option;
                return (
                    <option key={key} value={value}>
                        {key}
                    </option>
                );
            })}
        </select>
    );
};
