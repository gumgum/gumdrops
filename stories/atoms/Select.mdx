import { Canvas, Source } from '@storybook/blocks';
import { ArgsTable } from '@storybook/addon-docs';

import * as SelectStories from './Select.stories';
import Select from '../../components/atoms/Select';

# Select

The `<Select>` component is used to allow the user to choose a value from a set of options. It is recommended that you wrap it with the `<FormGroup>` component.

## Options

The default format for each option is `{ name: 'my name', value: 'my value' }`. If your options are in a different format, you may pass in `customValue` and/or `customName` props to override the default values of "name" and "value". For example, if your data looks like `{ id: 'my id', val: 'my value' }`, you can pass in `customName="id"` and `customValue="val"`.

## Controlled vs Uncontrolled

This component can be controlled or uncontrolled. This component accepts any prop that you may want to add. The only drawback is that there is not validation for unlisted props. **It is highly recommended that you use controlled components in React.**

_Recommended Props_: `name`, `value`, `defaultValue`, `onChange`

### Controlled example

```jsx
const fruitOptions = [
    { name: 'bananas', value: 1 },
    { name: 'apples', value: 2 }
];

const [selectedOption, setSelectedOption] = useState(1);
return (
    <FormGroup>
        <FormGroupLabel text="Fruits" />
        <Select
            options={fruitOptions}
            name="fruit"
            value={selectedOption}
            onChange={e => {
                setSelectedOption(e.target.value);
            }}
        />
    </FormGroup>
);
```

### Uncontrolled example

```jsx
_handleSubmit = ({ target }) => console.log(target.fruit.value);

render() {
    return (
        <form onSubmit={ this._handleSubmit }>
            <FormGroup>
                <FormGroupLabel text="Fruits" />
                <Select
                   options={ myOptions }
                   name="fruit"
                   defaultValue="3"
                />
            </FormGroup>
        </form>
    );
}
```

## Component Example

<Canvas of={SelectStories.Default} />

## Component Example Source

<Source of={SelectStories.Default} />

## Select Props

<ArgsTable of={Select} />
