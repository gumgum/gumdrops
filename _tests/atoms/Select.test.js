import renderer from 'react-test-renderer';
import React from 'react';
import Select from '../../components/atoms/Select';

const defaultProps = {
    context: 'success',
    size: 'lg',
    text: '100',
    options: [
        {
            value: 'foo',
            name: 'Foot'
        },
        {
            value: 'woo',
            name: 'Woot'
        }
    ]
};

test('Expect <Select> to render properly', () => {
    const tree = renderer.create(<Select {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
});
