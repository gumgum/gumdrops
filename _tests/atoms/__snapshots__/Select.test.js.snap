// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Select> to render properly 1`] = `
<select
  className="gds-form-group__select-input gds-form-group__select-input--success gds-form-group__select-input--lg"
  text="100"
>
  <option
    value="foo"
  >
    Foot
  </option>
  <option
    value="woo"
  >
    Woot
  </option>
</select>
`;
